examples_tests/60.ClusteredRendering/lighting.frag

```diff
@@ -0,0 +1,270 @@
+    const float spotHalfAngle = dot(spotDir,fromLight)*inversesqrt(dot(spotDir,spotDir)*dot(fromLight,fromLight));
+    const float spotPenumbra = clamp(((spotHalfAngle/cosineRange) - light.outerCosineOverCosineRange), 0.f, 1.f);
+
+    attenuation *= spotPenumbra;
+
+    const vec3 L = light.position.xyz-worldPos;
+    const float lenL2 = dot(L,L);
+    const float invLenL = inversesqrt(lenL2);
+    nbl_glsl_LightSample _sample = nbl_glsl_createLightSample(L*invLenL, interaction);
+
+    if (any(greaterThanEqual(intensity*attenuation,vec3(LIGHT_CONTRIBUTION_THRESHOLD))))
+        result += attenuation*intensity*nbl_bsdf_cos_eval(_sample,interaction, dUV);
+
+    return result;
+}
+
+vec3 nbl_computeLighting(out nbl_glsl_IsotropicViewSurfaceInteraction out_interaction, in mat2 dUV)
+{
+    const vec3 WorldPos = getWorldPosFromFramebufferCoords(gl_FragCoord);
+#ifdef CLIPMAP
+    const uint level = getClipmapLevel(WorldPos, EyePos);
+    const uint globalClusterID = getClipmapClusterAtLevel(level, WorldPos, EyePos);
+    ivec3 lightGridCoords = getLightGridCoords(globalClusterID);
+#endif
+#ifdef OCTREE
+    ivec3 lightGridCoords = getOctreeCluster(WorldPos, EyePos);
+#endif
+
+    nbl_glsl_IsotropicViewSurfaceInteraction interaction = nbl_glsl_calcSurfaceInteraction(
+        EyePos,
+        WorldPos,
+        Normal_WorldSpace,
+        mat2x3(dFdx(ViewPos),dFdy(ViewPos)));
+
+    
+    uint lightOffset, lightCount;
+    getLightIndexListOffsetAndCount(lightGridCoords, lightOffset, lightCount);
+
+
+    vec3 lightAccum = vec3(0.f);
+#if 1
+    for (uint i = 0u; i < lightCount; ++i)
+#else
+    for (uint lightID = 0u; lightID < LIGHT_COUNT; ++lightID)
+#endif
+    {
+#if 1
+        const uint lightID = texelFetch(lightIndexList, int(lightOffset + i)).x;
+#endif
+        const nbl_glsl_ext_ClusteredLighting_SpotLight light = ssbo.lights[lightID];
+
+        lightAccum += computeLightContribution(WorldPos, light, interaction, dUV);
+    }
+
+    nbl_glsl_MTLMaterialParameters mtParams = nbl_glsl_getMaterialParameters();
+#ifndef _NO_UV
+    if ((mtParams.extra&map_bump_MASK) == map_bump_MASK)
+    {
+        interaction.N = normalize(interaction.N);
+
+        vec2 dh = nbl_sample_bump(UV, dUV);
+
+        interaction.N = nbl_glsl_perturbNormal_derivativeMap(interaction.N, dh, interaction.V.dPosdScreen, dUV);
+    }
+#endif
+
+    vec3 Ka;
+    switch ((mtParams.extra&ILLUM_MODEL_MASK))
+    {
+    case 0:
+    {
+#ifndef _NO_UV
+    if ((mtParams.extra&(map_Kd_MASK)) == (map_Kd_MASK))
+        Ka = vec3(nbl_sample_bump(UV, dUV).rg, 0.f);
+    else
+#endif
+        Ka = mtParams.Kd;
+    }
+    break;
+    default:
+#define Ia 0.1
+    {
+#ifndef _NO_UV
+    if ((mtParams.extra&(map_Ka_MASK)) == (map_Ka_MASK))
+        Ka = nbl_sample_Ka(UV, dUV).rgb;
+    else
+#endif
+        Ka = mtParams.Ka;
+    Ka *= Ia;
+    }
+#undef Ia
+    break;
+    }
+
+    out_interaction = interaction;
+
+    return lightAccum + Ka;
```

Comment:
can you clean it up and just do calls to `nbl_glsl_bxdf_cos_eval` ? (so then it could override any material/pipeline)